docs/DafnyRef/Types.md

```diff
@@ -2294,9 +2295,230 @@ When `{:opaque}` is specified for function `g`, `g` is opaque,
+the non-`older` parameters.
+
+To understand what this means and why this attribute is useful,
+consider the following example, which specifies reachability between
+nodes in a directed graph. A `Node` is declared to have any number of
+children:
+
+```
+class Node {
+  var children: seq<Node>
+}
+```
+
+There are several ways one could specify reachability between
+nodes. One way (which is used in `Test/dafny1/SchorrWaite.dfy` in the
+Dafny test suite) is to define a type `Path`, representing lists of
+`Node`s, and to define a predicate that checks if a given list of
+`Node`s is indeed a path between two given nodes:
+
+```
+datatype Path = Empty | Extend(Path, Node)
+
+predicate ReachableVia(source: Node, p: Path, sink: Node, S: set<Node>)
+  reads S
+  decreases p
+{
+  match p
+  case Empty =>
+    source == sink
+  case Extend(prefix, n) =>
+    n in S && sink in n.children && ReachableVia(source, prefix, n, S)
+}
+```
+
+In a nutshell, the definition of `ReachableVia` says
+
+* An empty path lets `source` reach `sink` just when
+  `source` and `sink` are the same node.
+* A path `Extend(prefix, n)` lets `source` reach `sink` just when
+  the path `prefix` lets `source` reach `n` and `sink` is one of
+  the children nodes of `n`.
+
+To be admissible by Dafny, the recursive predicate must be shown to
+terminate. Termination is assured by the specification `decreases p`,
+since every such datatype value has a finite structure and every
+recursive call passes in a path that is structurally included in the
+previous. Predicate `ReachableVia` must also declare (an upper bound
+on) which heap objects that it depends on. For this purpose, the
+predicate takes an additional parameter `S`, which is used to limit
+the set of intermediate nodes in the path. More precisely, predicate
+`ReachableVia(source, p, sink, S)` returns `true` if and only if `p`
+is a list of nodes in `S` and `source` can reach `sink` via `p`.
+
+Using predicate `ReachableVia`, we can now define reachability in `S`:
+
+```
+predicate Reachable(source: Node, sink: Node, S: set<Node>)
+  reads S
+{
+  exists p :: ReachableVia(source, p, sink, S)
+}
+```
+
+This looks like a good definition of reachability, but Dafny won't
+admit it. The reason is twofold:
+
+* Quantifiers and comprehensions are allowed to range only over
+  allocated state. Ater all, Dafny is a type-safe language where every
+  object reference is _valid_ (that is, a pointer to allocated storage
+  of the right type)---it should not be possible, not even through a
+  bound variable in a quantifier or comprehension, for a program to
+  obtain an object reference that isn't valid.
+
+* This property is ensured by disallowing _open-ended_ quantifiers.
+  More precisely, the object references that a quantifier may range
+  over must be shown to be confined to object references that are
+  already allocated. Quantifiers that are not open-ended are called
```

Comment:
"already", what is the current time? I think it's not the time at which the predicate is called, because if I do two predicate calls with equivalent arguments, then the quantifier references must be allocated before the first call, even when evaluating the second call.

Shouldn't this say "More precisely, the object references that a quantifier may range
  over must be shown to be confined to object references that were allocated before one of the parameters passed to the predicate." ?